source/common/upstream/cluster_manager_impl.h

```diff
@@ -553,6 +666,17 @@ class ClusterManagerImpl : public ClusterManager, Logger::Loggable<Logger::Id::u
   using PendingUpdatesByPriorityMapPtr = std::unique_ptr<PendingUpdatesByPriorityMap>;
   using ClusterUpdatesMap = absl::node_hash_map<std::string, PendingUpdatesByPriorityMapPtr>;
 
+  /**
+   * Holds a reference to an on-demand CDS to keep it alive for the duration of a cluster discovery,
+   * and an expiration timer notifying worker threads about discovery timing out.
+   */
+  struct ClusterCreation {
+    OdCdsApiSharedPtr odcds_;
```

Comment:
Do we expect each on-demand route to really have a different CDS config source? I'm wondering how we can simplify the amount of bookkeeping and implementation complexity here if this is not needed.

Reply:
The `allocateOdCdsApi` function could do the deduplication based on the config and the resource locator. So if all the routes are using the same config, there would be just one shared instance of OdCdsApi. The deduplication is not yet implemented, though.